pinot-core/src/main/java/org/apache/pinot/core/udf/UdfParameter.java

```diff
@@ -0,0 +1,162 @@
+    _name = name;
+    _dataType = dataType;
+    _multivalued = multivalued;
+    _constraints = constraints;
+    _description = description;
+  }
+
+  public static UdfParameter result(FieldSpec.DataType dataType) {
+    return of("result", dataType);
+  }
+
+  public static UdfParameter of(String name, FieldSpec.DataType dataType) {
+    return new UdfParameter(name, dataType, false, List.of(), null);
+  }
+
+  public UdfParameter asMultiValued() {
+    if (_multivalued) {
+      return this;
+    }
+    return new UdfParameter(_name, _dataType, true, _constraints, _description);
+  }
+
+  public String getName() {
+    return _name;
+  }
+
+  public FieldSpec.DataType getDataType() {
+    return _dataType;
+  }
+
+  public boolean isMultivalued() {
+    return _multivalued;
+  }
+
+  public boolean isLiteralOnly() {
+    return _constraints.contains(LiteralConstraint.INSTANCE);
+  }
+
+  @Nullable
+  public String getDescription() {
+    return _description;
+  }
+
+  public UdfParameter withDescription(String description) {
+    return new UdfParameter(_name, _dataType, _multivalued, _constraints, description);
+  }
+
+  public UdfParameter asLiteralOnly() {
+    if (isLiteralOnly()) {
+      return this;
+    }
+    return new UdfParameter(_name, _dataType, _multivalued, List.of(LiteralConstraint.INSTANCE), _description);
+  }
+
+  /// Returns the list of constraints that this parameter has.
+  ///
+  /// This can be useful to define specific static constraints on the parameter, such as having a specific index.
+  public List<Constraint> getConstraints() {
+    return _constraints;
+  }
+
+  public UdfParameter constrainedWith(Constraint constraint) {
+    List<Constraint> newConstraints;
+    if (_constraints.isEmpty()) {
+      newConstraints = List.of(constraint);
+    } else {
+      newConstraints = new ArrayList<>(_constraints.size() + 1);
+      newConstraints.addAll(_constraints);
+      newConstraints.add(constraint);
+    }
+    return new UdfParameter(_name, _dataType, _multivalued, newConstraints, _description);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (!(o instanceof UdfParameter)) {
+      return false;
+    }
+    UdfParameter that = (UdfParameter) o;
+    return isMultivalued() == that.isMultivalued() && Objects.equals(_name, that._name)
+        && getDataType() == that.getDataType() && Objects.equals(getConstraints(), that.getConstraints())
+        && Objects.equals(getDescription(), that.getDescription());
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(_name, getDataType(), isMultivalued(), getConstraints(), getDescription());
+  }
+
+  @Override
+  public String toString() {
+    return _name + ": " + getTypeString();
+  }
+
+  public String getTypeString() {
+    String type;
+    if (isMultivalued()) {
+      type = "ARRAY(" + getDataType().name().toLowerCase() + ")";
+    } else {
+      type = getDataType().name().toLowerCase();
+    }
+    return type;
+  }
+
```

Comment:
nit: constraint implementations need to either use fixed set of values or implement equals() for UdfParameter.equals() to work correctly.